### PR TITLE
Allow dashes in author name #42

### LINKED ISSUE #42: Allow dashes in author name
e.g `//@@author john-reused`
The current code strips out any non alpha-numerics from the author name

### PR DESCRIPTION
**Pull request by yamgent**
Fixes #42.

Ready for review.

### CODE REVIEW

**Comment by riwu** [comment]
Should leading/trailing dashes in names be removed? eg. "--author-name -" to "author-name"
If so, this would do the job: https://github.com/se-edu/collate/compare/master...riwu:dashInName

**Comment by damithc** [comment]
@sebastianquek for your review please.

**Comment by sebastianquek** [comment]
**Thanks @yamgent for your contributions!**

### Background
Looking at the various use cases, the author tag can be followed by several variants of author names:

1. `// @@author A0123456Z`
2. `// @@author generated`
3. `// @@author A0123456Z-unused`
4. `// @@author A0123456Z-reused`
5. `<!-- @@author A0123456Z -->`
6. `<!-- @@author A0123456Z-unused -->`
7. `/* @@author A0123456Z */`

The current implementation allows for spaces in names, and when we allow dashes as well, we will need to handle edge cases such as in point 6 (as @yamgent has found). However, the fix is very specific to `-->`, and may be adding unnecessary complexity to the codebase.

Given that we currently restrict author names to only alphanumeric characters and dashes as in the above examples, I propose that we remove the feature that allows for spaces in names.

### Proposed implementation
One approach is to extract the author's name according to a regex expression. In a very verbose manner: `^[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9]`.

- The `^` indicates the beginning of the sentence,
- the 1st matching character set ensures the author's name has to start with an alphanumeric character, 
- the 2nd set allows any number of alphanumeric characters and `-`, 
- the 3rd set is similar to the 1st set.

This ensures that author names that start and/or end with non-alphanumeric characters are illegal (as @riwu has mentioned). Furthermore, alphanumeric characters can be interleaved with `-` characters (e.g. `long-author-name`).

- - -
Any thoughts/ideas?

**Comment by damithc** [comment]
I'm ok with not allowing spaces in author name.

**Review by sebastianquek — CHANGES_REQUESTED** [review]
Remove ability to have whitespace in author names, use regex that accepts and extracts the valid author names.

This streamlines the extraction of the author name from an author tag as special rules (HTML comments) are not required anymore.

File: src/main/java/backend/Logic.java [hunk 1]
```diff
@@ -32,14 +34,13 @@
...
     private static final String MARKDOWN_CODE_LANGUAGE_END = "```";
-    
-    private static final String HTML_COMMENT_END = "-->";
 
     private static final String STRING_EMPTY = "";
     private static final char STRING_PERIOD = '.';
 
-    private static final String REGEX_NEITHER_ALPHANUMERIC_NOR_SPACE_NOR_DASH =
-        "[^ a-zA-Z0-9\\-]+";
+    private static final String REGEX_AUTHOR_NAME_FORMAT = "(^[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9]).*";
```
Comment (sebastianquek) [inline src/main/java/backend/Logic.java]: The `.*` at the end will match any number of characters that are not line breaks! Do test it with the new test file you created (`testFileHtml.html`).

**Comment by sebastianquek** [comment]
One last thing, @yamgent could you add test cases for author names that start and/or end with `-`?

**Review by sebastianquek — APPROVED** [review]

**Comment by yamgent** [comment]
@sebastianquek Okay sure, am working on it.

File: src/main/java/backend/Logic.java [hunk 2]
```diff
@@ -36,8 +38,9 @@
     private static final String STRING_EMPTY = "";
     private static final char STRING_PERIOD = '.';
 
-    private static final String REGEX_NEITHER_ALPHANUMERIC_NOR_SPACE =
-        "[^ a-zA-Z0-9]+";
+    private static final String REGEX_AUTHOR_NAME_FORMAT = "(^[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9])";
```
Comment (riwu) [inline src/main/java/backend/Logic.java]: Is it intended to fail if author string doesn't start with alphanumeric? 
Removing '^' from the regex would simply strip leading non-alphanumeric instead of failing.

**Comment by yamgent** [comment]
@sebastianquek I added test cases for the dashes, do take a look.

@riwu thanks for your input!

File: src/main/java/backend/Logic.java (hunk 2 above)
Comment (sebastianquek) [reply src/main/java/backend/Logic.java]: You're right, if we allow `@@author matric-num----`, which results with the author name being `matric-num`,  `@@author --matric-num----` and `@@author !@%$&*^matric-num` likewise should not fail.

Comment (yamgent) [reply src/main/java/backend/Logic.java]: I fixed it in 769d213.

**Comment by sebastianquek** [comment]
Awesome! 👍
As an exercise about equivalence partitions and their associated boundary values, you can also try adding additional test cases. Recall that any number of non-alphanumeric characters before and after the author's name should not cause a failure. 😄

**Comment by yamgent** [comment]
@sebastianquek Few more test cases added

**Comment by sebastianquek** [comment]
LGTM 👍
